### ISSUE TITLE
Query with relativePath should return unique result

### ISSUE THREAD

**Issue by stefanprobst** [gatsbyjs/gatsby#10993]
## Description

A GraphQL query for files with `relativePath` currently is not unique, since `relativePath` refers to the path of the source instance as the basepath, and there can be multiple instances of `gatsby-source-filesystem`. This is confusing and can lead to surprising results.

### Steps to reproduce

See [this repo](https://github.com/stefanprobst/gatsby-relativepath), which has two instances of `gatsby-source-filesystem`:
```js
{
  resolve: `gatsby-source-filesystem`,
  options: {
    name: `feat1`,
    path: `${__dirname}/src/feat1`,
  },
},
{
  resolve: `gatsby-source-filesystem`,
  options: {
    name: `feat2`,
    path: `${__dirname}/src/feat2`,
  },
},
```

And this folder structure:
`src/feat1/images/gatsby-astronaut.png`
`src/feat2/images/gatsby-astronaut.png`

Now, [with a query like this](https://github.com/stefanprobst/gatsby-relativepath/blob/52fd39225b0ab9a5e4c6eddabe01b4852c69c229/src/components/image.js#L20-L24)
```graphql
query { file(relativePath: { eq: "images/gatsby-astronaut.png" } ) { id }
```
it is not clear which image you will get.

### Expected result

I think `relativePath` in a query should take the `componentPath` as the base path instead. This would be a breaking change.

**Comment by sidharthachatterjee** [comment]
Hey @stefanprobst 

This is a fair point. On a side note, you _could_ use sourceInstanceName as a filter along with relativePath
```
query {
  file(relativePath: {
    eq: "gatsby-astronaut.png"
  }, sourceInstanceName: {
    eq: "images"
  }) {
    id
    relativePath
    sourceInstanceName
  }
}
```

**Comment by stefanprobst** [comment]
True, but then you'd already have to be aware that the query result was not the expected one. The issue is more that the query might return unintended results without being immediately obvious.

**Comment by stefanprobst** [comment]
not stale

**Comment by KyleAMathews** [comment]
I'm not sure how we'd fix this? `relativePath` is inherently _not_ unique giving that any number of instances of gatsby-source-filesystem can create `File` nodes.

Do you have any ideas?

**Comment by stefanprobst** [comment]
Just thinking out loud: One way could be to make `relativePath` in a query relative to the component path, i.e. when we encounter `relativePath` in query args, translate that into an absolute path, with the `componentPath` as the base path. Something similar to what I did here (the commented out version): https://github.com/stefanprobst/gatsby/blob/69d759a4e6fd1f24a8dda05e0f5706766723bd95/packages/gatsby/src/schema/resolvers/special-cases.js#L30
I haven't tried this with fragments though. And another potential downside is that you'd have to adjust the query paths when you move the component somewhere else.

**Comment by pieh** [comment]
I'm just thinking that is more of a documentation issue. `relativePath` have been like this for a long time but it's not documented - neither in `gatsby-source-filesystem` README/docs nor in graphql schema (we could use `setFieldsOnGraphQLNodeType` right now to provide field description).

There is also this PR https://github.com/gatsbyjs/gatsby/pull/10540 than really changes what `relative(Path|Directory)` is (as side effect of other change rather than main point) - which would make those fields unique.

**Comment by jnsrikanth** [comment]
has this been resolved? Does the PR #10540 really address this?

**Comment by LekoArts** [comment]
Closing this as stale since in the meantime Gatsby v3 and updated related packages were released. Please try with the latest versions and if you still see this problem open a new bug report (it must include a [minimal reproduction][reproduction]).

[reproduction]: https://gatsby.dev/reproduction
